### DISCUSSION TITLE
question on preventing css flickering when implementing dark mode

### DISCUSSION THREAD

**Discussion by aquaductape** [vercel/next.js#12533] [Help] [answered]
I have a simple setup that persists a theme(such as dark mode) by using css variables and localStorage. My goal is to load the page with the correct theme, without the flashing of the default css theme(in this case it would be a light theme).

Here's an example below that shows flashing as I refresh. The flashing is more prevalent upon loading routes.
![flashing](https://user-images.githubusercontent.com/29286430/81105330-a28d6400-8ec8-11ea-80ae-6ba5ed30b89d.gif)

The way I prevented flashing was running a script before the content loads, to read local storage, get theme, and set css variables to the appropriate theme.

`_document.js`
```js
import Document, { Html, Head, Main, NextScript } from "next/document";

class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head>
          <script src="/static/theme.js"></script>
        </Head>
        <body>
          {/* some sources placed script inside body, before Main, but I experienced flashing */}
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

export default MyDocument;
```
`/static/theme.js`
```js
(function () {
  var currentTheme;

  function changeTheme(inputTheme) {
    if (inputTheme === "dark") {
      const theme = themeConfig.dark;
      for (let key in theme) {
        setCSSVar(key, theme[key]);
      }
      localStorage.setItem("theme", inputTheme);
    } else {
      const theme = themeConfig.light;
      for (let key in theme) {
        setCSSVar(key, theme[key]);
      }
      localStorage.setItem("theme", inputTheme);
    }
  }

  function setCSSVar(property, color) {
    document.documentElement.style.setProperty(property, color);
  }
  try {
    currentTheme = localStorage.getItem("theme") || "light";

    var themeConfig = {
      dark: {
        "--color-homepage-bg": "#0e141b",
        "--color-text": "#fff",
        "--code": "#ccc",
        "--color-blue": "#f300e0",
        "--color-grey": "#ccc",
      },
      light: {
        "--color-homepage-bg": "#fff",
        "--color-text": "#000",
        "--code": "#f1f1f1",
        "--color-blue": "#0070f3",
        "--color-grey": "#eaeaea",
      },
    };

    changeTheme(currentTheme);
  } catch (err) {
    console.log(new Error("accessing theme has been denied"));
  }
})();
```
Here's the result with no flashing, as I refresh. [Live Demo](https://nextjs-toggle-theme-nonflash-b8cko6a71.now.sh/)
![no-flashing](https://user-images.githubusercontent.com/29286430/81107101-7a533480-8ecb-11ea-9999-1285b4c5978d.gif)

Even though I solved my problem, is it optimal or a bad solution? Here's the [source code](https://github.com/aquaductape/nextjs-toggle-theme-nonflash).

**Comment by rafaelalmeidatk** [comment, marked as answer]
Yeah this is a good solution, Dan did the same with his blog: https://github.com/gaearon/overreacted.io/blob/master/src/html.js#L21

But I would inline the JS like he did instead of making a request for it.

**Reply by rafaelalmeidatk** [reply]
Btw, this is unrelated for the question but the `static` folder has been deprecated in favor of the [`public` folder](https://nextjs.org/docs/basic-features/static-file-serving)

**Reply by aquaductape** [reply]
Interesting, I didn't know that `static` was a special directory, I named it that way because on my main project, I was working with typescript. Since the script is a JS file I placed it in a different folder, which I named `static`.

Placing the script file in `public` has issues when navigating to routes through url address. Loading home page it finds the script, but when loading to other pages it tries to find the script on that pathname rather than getting it at root of `public`.
For example, I'll open `http://localhost:3000/docs/fish` which results in `GET http://localhost:3000/docs/theme.js net::ERR_ABORTED 404 (Not Found)`

Previous attempt to run script by accessing it from public
`_document.js`
```js
class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head>
          <script src="theme.js"></script>
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

export default MyDocument;
```

But that doesn't matter since I'm switching to inline JS in script tag that you recommended, which by the way, thanks :+1: 😁

**Reply by jason-shen** [reply]
i follow your code but still getting the flash very strange one here

**Reply by aquaductape** [reply]
@jason-shen Mind sharing the a reproducible example?

**Comment by brianlovin** [comment]
Yep, this looks good - basically the same thing that https://joshwcomeau.com/gatsby/dark-mode/ found as well, but as @rafaelalmeidatk mentions, inlining the JS is probably more correct.

**Reply by HiranmayaGundu** [reply]
Hey @brianlovin! I'm trying to do the same thing that was suggested in the blog above.  I had to replace `Object.entries().forEach` with a `for (key in object)`  because it seems like `Object.entries` was automatically being pollyfilled. Is there any way to prevent the pollyfilling?

**Reply by aquaductape** [reply]
I'd suggest you stick with `for in..` loop instead of `Object.entries` [because of speed](https://hackernoon.com/5-techniques-to-iterate-over-javascript-object-entries-and-their-performance-6602dcb708a8). In this specific case I would value speed over readability since we want to the script to finish the dark mode logic as fast as possible since it's blocking the rendering of the page content. 

Your theme could easily go over 50 css variables, stack overflow has around 100, facebook has 350.

**Reply by HiranmayaGundu** [reply]
Ah, did not know about the performance characteristics of the various iteration techniques. Thanks @aquaductape!

**Comment by kachkaev** [comment]
I’ve used a similar trick in https://njt.now.sh, here is the commit: https://github.com/kachkaev/njt/commit/1ef3bb026b006a355173c8a55594de65de53c102

Inlining JS is quicker than making a separate network request. To further reduce the network payload a bit, it is possible to compress inline JS with [terser](https://www.npmjs.com/package/terser):

```tsx
// in _document.tsx:
import Terser from "terser";
import mem from "mem";

const minify = mem(Terser.minify);

const InlineJs: React.FunctionComponent<{ code: string; children?: never }> = ({
  code,
}) => {
  const minifyOutput = minify(code);

  if (minifyOutput.error) {
    throw minifyOutput.error;
  }

  if (!minifyOutput.code) {
    throw new Error("Minified code is empty");
  }

  return <script dangerouslySetInnerHTML={{ __html: minifyOutput.code }} />;
}

// ...

<InlineJs code={`...`} />
```

Also, thanks to a hook from the [use-dark-mode](https://github.com/donavon/use-dark-mode) package, I got auto-syncing with the macOS theme without page reload 👀

 
![dark mode small](https://user-images.githubusercontent.com/608862/82753120-30b78480-9dbb-11ea-8613-7867b7e3e7e4.gif)

**Comment by ThanosDodd** [comment]
Excellent solution.<br/>Has anyone updated it for Next 13's app directory? Because it seems only next/script is accepted for one thing, plus when using Typescript you can't export a self-executing function without `export {}` which paradoxically returns a `Uncaught SyntaxError: Unexpected token 'export'` error<br/>Also it won't execute anyway if you add it in the Head instead of inside the body

**Comment by jamesvclements** [comment]
Wondering if this still works for everyone in Next.js 13.
My injected script runs locally, but it doesn't seem to execute when deployed. 

```jsx
<body>
  {/* The script here needs to run before the rest of the body to prevent default theme flicker */}
  <MagicScriptTag />
  <ThemeProvider>
    <StoreProvider>
```

```jsx
const MagicScriptTag = () => {
  console.log("in magic script tag...");
  const boundFn = String(setColorsByTheme).replace("✸", JSON.stringify(COLORS));
  let calledFunction = `(${boundFn})()`;
  // eslint-disable-next-line react/no-danger
  return <script dangerouslySetInnerHTML={{ __html: calledFunction }} />;
};

```

**Comment by hungtcs** [comment]
I use a similar way in nextjs 13, It works but I need to use attribute instead of css variables,
So when I refresh the page, I get the following warning:
```
Warning: Extra attributes from the server: data-color-mode
```

```ts
const memoizedMinify = mem(minify);

export default async function RootLayout({ children }: PropsWithChildren) {
  const setColorModeCode = await memoizedMinify(`
    (function() {
      if (window.localStorage.getItem('rhino.color_mode') === 'dark' || window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.documentElement.setAttribute('data-color-mode', 'dark');
      }
    }());
  `);

  return (
    <html lang="en">
      <head>
        <script dangerouslySetInnerHTML={{__html: setColorModeCode.code!}}></script>
      </head>
      <body className={materialSymbolsOutlinedFont.variable}>
        {children}
      </body>
    </html>
  )
}
```

**Comment by Arndeverafter** [comment]
One other way to accomplish the same and prevent flickering with a pure JS implementation would be to inline the following

```
  (function () {
    var currentTheme = "light";

    if (localStorage.getItem("theme")) {
      currentTheme = localStorage.getItem("theme");
    } else if (
      window.matchMedia &&
      window.matchMedia("(prefers-color-scheme: dark)").matches
    ) {
      currentTheme = "dark";
    }

    if (currentTheme === "dark") {
      document.documentElement.classList.add("dark");
    } else {
      document.documentElement.classList.remove("dark");
    }
    document.documentElement.setAttribute("data-theme", currentTheme);
    localStorage.setItem("theme", currentTheme);
  })();
```

**Reply by harshal255** [reply]
> One other way to accomplish the same and prevent flickering with a pure JS implementation would be to inline the following
> 
> ```
>   (function () {
>     var currentTheme = "light";
> 
>     if (localStorage.getItem("theme")) {
>       currentTheme = localStorage.getItem("theme");
>     } else if (
>       window.matchMedia &&
>       window.matchMedia("(prefers-color-scheme: dark)").matches
>     ) {
>       currentTheme = "dark";
>     }
> 
>     if (currentTheme === "dark") {
>       document.documentElement.classList.add("dark");
>     } else {
>       document.documentElement.classList.remove("dark");
>     }
>     document.documentElement.setAttribute("data-theme", currentTheme);
>     localStorage.setItem("theme", currentTheme);
>   })();
> ```

Thankyou 🥂

**Comment by bloodykheeng** [comment]
@aquaductape 
just use cookies 

see my layout.tsx
so i get the default theme through a cookie i set on from client but it persists or syncs on server

```
import { ThemeProvider } from '@/util/ThemeProvider'


export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {

  const getCookie = async () => {
    const cookieStore = await cookies()
    return cookieStore.get("theme")
  }

  const defaultTheme = getCookie();

  return (
    <html lang="en" suppressHydrationWarning>
      <head>
        <link id="theme-link" rel="stylesheet" href="/themes/lara-light-blue/theme.css" />


      </head>
      <body className={`bg-[#FCFCFC] dark:bg-black ${inter.className}`}>

        <NextJsProgressBar />

        <TanstackProvider>
       
            <ThemeProvider defaultTheme={defaultTheme}>
              <PrimeReactProvider>
                <PrimeReactThemeSync>
                  <ToastProvider>
                    <PrimeReactToastProvider>
                      <AuthProvider>
                        <FirebaseNotificationProvider>

                          {children}

                        </FirebaseNotificationProvider>
                      </AuthProvider>
                    </PrimeReactToastProvider>
                  </ToastProvider>
                </PrimeReactThemeSync>
              </PrimeReactProvider>
            </ThemeProvider>
 
        </TanstackProvider>

      </body>
    </html>
  );
}


```



there is my custom theme provider

```
"use client";

import { createContext, useState, useEffect, useContext } from "react";
import Cookies from "js-cookie";

type Theme = "light" | "dark";
type ThemeContextType = {
    theme: Theme;
    setTheme: (theme: Theme) => void;
};

const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

export function ThemeProvider({ children, defaultTheme }: { children: React.ReactNode, defaultTheme: any }) {

    const savedTheme = Cookies.get("theme") as Theme;
    const [theme, setTheme] = useState<Theme>(defaultTheme ?? savedTheme ?? "light");

    useEffect(() => {
        const html = document.documentElement;
        if (savedTheme) {
            setTheme(savedTheme);
        } else {
            Cookies.set("theme", theme, { expires: 365 }); // set initial default theme
        }

        if (savedTheme === "dark") {
            html.classList.add("dark");
            html.classList.remove("light");
            html.style.colorScheme = "dark";
            html.setAttribute("data-theme", 'dark'); // Set data-theme attribute
        } else {
            html.classList.add("light");
            html.classList.remove("dark");
            html.style.colorScheme = "light";
            html.setAttribute("data-theme", 'light'); // Set data-theme attribute
        }
    }, []);

    const handleSetTheme = (newTheme: Theme) => {
        setTheme(newTheme);
        Cookies.set("theme", newTheme, { expires: 365 }); // Store for a year

        const html = document.documentElement;

        if (newTheme === "dark") {
            html.classList.add("dark");
            html.classList.remove("light");
            html.style.colorScheme = "dark";
            html.setAttribute("data-theme", 'dark'); // Set data-theme attribute
        } else {
            html.classList.add("light");
            html.classList.remove("dark");
            html.style.colorScheme = "light";
            html.setAttribute("data-theme", 'light'); // Set data-theme attribute
        }
    };

    return (
        <ThemeContext.Provider value={{ theme, setTheme: handleSetTheme }}>
            {children}
        </ThemeContext.Provider>
    );
}

export function useTheme() {
    const context = useContext(ThemeContext);
    if (!context) {
        throw new Error("useTheme must be used within a ThemeProvider");
    }
    return context;
}

```



usage 
how i use my theme
i stopped using next-themes it uses local storage so its not effective thats why i resorted to a custom hook

```
"use client";


import { Button } from "primereact/button";
import { PrimeIcons } from "primereact/api";

// import { useTheme } from "next-themes";
import { useTheme } from "@/util/ThemeProvider";


const ThemeToggler = () => {
  const { theme, setTheme } = useTheme();


  return (
    <Button
      icon={theme === "dark" ? PrimeIcons.SUN : PrimeIcons.MOON}
      className="p-button-rounded p-button-text mt-1 mb-1  text-black dark:text-white"
      onClick={() => setTheme(theme === "dark" ? "light" : "dark")}
      aria-label="theme toggler"
    />
  );
};

export default ThemeToggler;
```
